### PR TITLE
Add Bitcoin ingester

### LINKED ISSUE #20: Bitcoin Ingester: deploy to public production cluster
We need to deploy a Bitcoin Ingester to the open-source/public production cluster. Consult with @kaladinlight prior to starting ticket.

### LINKED ISSUE #30: Bitcoin Ingester: Get blocks from node
Get new blocks from the node and put them in the txid queue

### PR DESCRIPTION
**Pull request by mvpratt**
New blocks:
- subscribe to new blocks from the node and put in the block queue
- parse transactions from block and put in the txid queue
- parse addresses from each transaction and check the registry.

Mempool:
- subscribe to new transactions from mempool and put in the txid queue

### CODE REVIEW

File: coinstacks/bitcoin/ingester/src/topology.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+const BROKER_URL = process.env.BROKER_URL as string
+const COINSTACK = process.env.COINSTACK as string
+
+if (!BROKER_URL) throw new Error('BROKER_URL env var not set')
+if (!COINSTACK) throw new Error('COINSTACK env var not set')
+
+const connection = new Connection(BROKER_URL)
+const deadLetterExchange = `exchange.${COINSTACK}.deadLetter`
+
+const topology: Connection.Topology = {
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/topology.ts:13]: @kaladinlight what do you think of using a COINSTACK env to generate coin-specific message queues? (vs hard-coding)

Comment (kaladinlight) [reply coinstacks/bitcoin/ingester/src/topology.ts:13]: Could potentially be used for whatever base set of common exchanges/queues we deem necessary. I could see the possibility where a specific coinstack might need some extra queues for processing unique data set maybe? Then there is the even bigger discussion of can we replace rabbit with something lighter weight to provide similar functionality. I am in camp copy for now until we actually get more than a couple coins or only blockbook coins. Would be easy enough to abstract later if we want.

Comment (mvpratt) [reply coinstacks/bitcoin/ingester/src/topology.ts:13]: so hardcode to bitcoin then?

File: coinstacks/bitcoin/ingester/package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  },
+  "dependencies": {
+    "@shapeshiftoss/blockbook": "^0.1.1",
+    "@shapeshiftoss/common-ingester": "^0.1.1",
+    "@shapeshiftoss/common-mongo": "^0.1.1",
+    "@shapeshiftoss/logger": "^0.1.1",
+    "@shapeshiftoss/thorchain": "^0.1.1",
+    "axios": "^0.19.2",
+    "bignumber.js": "^9.0.1",
+    "ethers": "^5.4.0"
```
Comment (kaladinlight) [inline coinstacks/bitcoin/ingester/package.json:24]:
```suggestion
```

File: coinstacks/bitcoin/ingester/package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "build": "tsc",
+    "clean": "rm -rf dist node_modules"
+  },
+  "dependencies": {
+    "@shapeshiftoss/blockbook": "^0.1.1",
+    "@shapeshiftoss/common-ingester": "^0.1.1",
+    "@shapeshiftoss/common-mongo": "^0.1.1",
+    "@shapeshiftoss/logger": "^0.1.1",
+    "@shapeshiftoss/thorchain": "^0.1.1",
+    "axios": "^0.19.2",
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/package.json:22]: rm big number and ethers

File: coinstacks/bitcoin/ingester/src/topology.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    { name: `queue.bitcoin.tx.unchained`, options: { durable: true } }, // default unchained client queue for development
+    { name: `queue.bitcoin.registry.deadLetter`, options: { durable: true } },
+    { name: `queue.bitcoin.newBlock.deadLetter`, options: { durable: true } },
+    { name: `queue.bitcoin.block.deadLetter`, options: { durable: true } },
+    { name: `queue.bitcoin.txid.deadLetter`, options: { durable: true } },
+    { name: `queue.bitcoin.txid.address.deadLetter`, options: { durable: true } },
+    { name: `queue.bitcoin.tx.deadLetter`, options: { durable: true } },
+  ],
+  bindings: [
+    { source: 'exchange.unchained', queue: `queue.bitcoin.registry`, pattern: 'bitcoin`.registry' },
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/topology.ts:39]:
```suggestion
    { source: 'exchange.unchained', queue: `queue.bitcoin.registry`, pattern: 'bitcoin.registry' },
```

File: coinstacks/bitcoin/ingester/src/topology.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    { source: `exchange.bitcoin`, queue: `queue.bitcoin.newBlock`, pattern: 'newBlock' },
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/topology.ts:40]: remove back ticks ```

File: coinstacks/bitcoin/ingester/src/types.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  hash: string
+  confirmations: number
+  size: number
+  strippedsize: number
+  weight: number
+  height: number
+  version: number
+  versionHex: string
+  merkleroot: string
+  tx: Array<string> /// todo
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/types.ts:14]:
```suggestion
  tx: Array<string>
```

File: coinstacks/bitcoin/ingester/src/types.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  mediantime: number
+  nonce: number
+  bits: string
+  difficulty: number
+  chainwork: string
+  nTx: number
+  previousblockhash: string
+  nextblockhash: string
+}
+export interface ReorgResult {
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/types.ts:25]:
```suggestion

export interface ReorgResult {
```

File: coinstacks/bitcoin/ingester/src/workers/newBlock.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import { BTCBlock, ReorgResult } from '../types'
+
+const NODE_ENV = process.env.NODE_ENV
+const RPC_URL = process.env.RPC_URL as string
+
+if (NODE_ENV !== 'test') {
+  if (!RPC_URL) throw new Error('RPC_URL env var not set')
+}
+
+const REORG_BUFFER = 1
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/newBlock.ts:15]:
```suggestion
const REORG_BUFFER = 6
```

File: coinstacks/bitcoin/ingester/src/workers/newBlock.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    await blocks.getByHash(dbBlock.prevHash),
+    await getBlockByHash(nodeBlock.previousblockhash)
+  )
+}
+
+const onMessage = (newBlockWorker: Worker, reorgWorker: Worker) => async (message: Message) => {
+  const newBlock: NewBlock = message.getContent()
+
+  try {
+    await processBlock(newBlockWorker, reorgWorker)
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/newBlock.ts:126]: rm processBlock fn

File: coinstacks/bitcoin/ingester/src/workers/newBlock.ts [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+const processBlock = async (newBlockWorker: Worker, reorgWorker: Worker): Promise<boolean> => {
+  let dbBlockLatest = await blocks.getLatest()
+  logger.debug('dbBlockLatest:', dbBlockLatest?.height)
+
+  const nodeHeight = await getHeight()
+  logger.debug('nodeHeight:', nodeHeight)
+
+  let height = dbBlockLatest ? dbBlockLatest.height + 1 : nodeHeight - REORG_BUFFER
+  while (height <= nodeHeight) {
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/newBlock.ts:65]: change back to hashorheight like eth

File: coinstacks/bitcoin/ingester/src/workers/newBlock.ts [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    queueName: `queue.bitcoin.newBlock`,
+    exchangeName: `exchange.bitcoin.block`,
+  })
+
+  const reorgWorker = await Worker.init({
+    exchangeName: `exchange.bitcoin`,
+  })
+
+  // prime the pump
+  await processBlock(newBlockWorker, reorgWorker)
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/newBlock.ts:145]: remove this and verify that fetches blocks when db is empty

File: coinstacks/bitcoin/ingester/src/workers/block.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { Worker, Message } from '@shapeshiftoss/common-ingester'
+import { logger } from '@shapeshiftoss/logger'
+import { BTCBlock } from '../types'
+
+const onMessage = (worker: Worker) => (message: Message) => {
+  const block: BTCBlock = message.getContent()
+  logger.debug(`block: (${Number(block.height)}) ${block.hash}`)
+
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/block.ts:8]: remove - process all txs

File: coinstacks/bitcoin/ingester/src/workers/block.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+const main = async () => {
+  const worker = await Worker.init({
+    queueName: `queue.bitcoin.block`,
+    exchangeName: `exchange.bitcoin.txid`,
+  })
+
+  worker.queue?.prefetch(1)
+  logger.info('block.worker.queue.activateConsumer()')
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/block.ts:25]:
```suggestion
```

File: coinstacks/bitcoin/ingester/src/workers/txid.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+if (!INDEXER_URL) throw new Error('INDEXER_URL env var not set')
+
+const blockbook = new Blockbook(INDEXER_URL)
+
+const onMessage = (worker: Worker) => async (message: Message) => {
+  const txid: string = message.getContent()
+
+  try {
+    const tx = await blockbook.getTransaction(txid)
+    logger.debug(`txid: ${txid}`)
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/txid.ts:16]:
```suggestion
```

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+  return pages
+}
+
+// todo - can there be more than one address per vin?
+const getAddresses = (tx: Tx): Array<string> => {
+  let addresses: Array<string> = []
+
+  const sendAddresses = tx.vin.map((vin) => (vin.addresses ? vin.addresses[0] : ''))
+  const receiveAddresses = tx.vin.map((vin) => (vin.addresses ? vin.addresses[0] : ''))
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:40]:
```suggestion
  const receiveAddresses = tx.vout.map((vout) => (vout.addresses ? vout.addresses[0] : ''))
```

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+// note - assume exactly one address per vin
+const getAddresses = (tx: Tx): Array<string> => {
+  let addresses: Array<string> = []
+
+  const sendAddresses = tx.vin.map((vin) => (vin.addresses ? vin.addresses[0] : ''))
+  const receiveAddresses = tx.vout.map((vout) => (vout.addresses ? vout.addresses[0] : ''))
+  addresses = addresses.concat(sendAddresses, receiveAddresses)
+  const addressesFiltered = addresses.filter((address) => address != '')
+
+  logger.debug(`addresses: ${addresses}`)
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:44]:
```suggestion
```

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+  }
+
+  return pages
+}
+
+// note - assume exactly one address per vin
+const getAddresses = (tx: Tx): Array<string> => {
+  let addresses: Array<string> = []
+
+  const sendAddresses = tx.vin.map((vin) => (vin.addresses ? vin.addresses[0] : ''))
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:39]: are there multiple addresses for multi-sig?

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+    pages.push(page)
+  }
+
+  return pages
+}
+
+// note - assume exactly one address per vin
+const getAddresses = (tx: Tx): Array<string> => {
+  let addresses: Array<string> = []
+
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:38]: consider forEach vs map

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 19]
```diff
@@ -0,0 +1,177 @@
+import { Tx, Blockbook } from '@shapeshiftoss/blockbook'
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:1]: remove delta sync?

File: coinstacks/bitcoin/ingester/src/workers/tx.ts [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+  let addresses: Array<string> = []
+
+  const sendAddresses = tx.vin.map((vin) => (vin.addresses ? vin.addresses[0] : ''))
+  const receiveAddresses = tx.vout.map((vout) => (vout.addresses ? vout.addresses[0] : ''))
+  addresses = addresses.concat(sendAddresses, receiveAddresses)
+  const addressesFiltered = addresses.filter((address) => address != '')
+
+  logger.debug(`addresses: ${addresses}`)
+  return addressesFiltered
+}
```
Comment (mvpratt) [inline coinstacks/bitcoin/ingester/src/workers/tx.ts:46]: 1. remove duplicate addresses   `return [...new Set(addresses.map((address) => utils.getAddress(address)))]`
2. check into bitcoin addresses checksums (base58 check?)

File: coinstacks/bitcoin/ingester/src/workers/tx.ts (hunk 17 above)
Comment (mvpratt) [reply coinstacks/bitcoin/ingester/src/workers/tx.ts:39]: multi-sig are still a single address which is the script hash

File: coinstacks/bitcoin/ingester/src/workers/tx.ts (hunk 20 above)
Comment (mvpratt) [reply coinstacks/bitcoin/ingester/src/workers/tx.ts:45]: all bitcoin addrs are base58.  there's no non-checksum address format like ethereum

File: coinstacks/bitcoin/ingester/src/workers/tx.ts (hunk 19 above)
Comment (mvpratt) [reply coinstacks/bitcoin/ingester/src/workers/tx.ts:1]: done

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**